skema/text_reading/text_reading/src/main/scala/org/ml4ai/skema/text_reading/scenario_context/CosmosOrderer.scala

```diff
@@ -0,0 +1,51 @@
+package org.ml4ai.skema.text_reading.scenario_context
+import org.apache.poi.openxml4j.exceptions.InvalidOperationException
+import org.clulab.odin.Mention
+import org.ml4ai.skema.text_reading.attachments.MentionLocationAttachment
+
+/**
+  * Implementation to resolve ordering for mentions that come from a Cosmos Json file.
+  * Will use the `MentionLocationAttachment` to inform the final linear order using page numbers
+  *
+  * @param ms All the mentions from the Cosmos Json, used to establish the order bounds
+  */
+class CosmosOrderer(ms:Iterable[Mention]) extends MentionOrderer {
+
+  private def getMentionLocation(m:Mention) = m.attachments.filter(_.isInstanceOf[MentionLocationAttachment]).head.asInstanceOf[MentionLocationAttachment]
+
+  private case class OrderingHelper(page:Int, block:Int, sent:Int, s:Int  = 0)
+
+  // Will use the `MentionLocationAttachment` on each mention to generate an offset for each (page, block)
+  private val offsets =
+    ms.toSeq.map(
+      m => {
+        if (!m.attachments.exists(_.isInstanceOf[MentionLocationAttachment]))
+          throw new InvalidOperationException("Mention without MentionLocationAttachment")
+
+        val location = getMentionLocation(m)
+        val page = location.pageNum.head
+        val block = location.blockIdx.head
+        val numSentences = m.document.sentences.length
+        OrderingHelper(page, block, numSentences)
+      }
+    ).distinct.sortBy{
+      case OrderingHelper(p,b,s, _) => (p, b, s)
```

Comment:
I'm not sure why numSentences should be part of the sort, but then I don't understand the data structure.  If there are two Mentions on the same page and block, then the one whose document has the most sentences should come first?

Reply:
Num sentences is just carried on to subsequent functional calls. On the case of several mentions, its value would be equal so it doesn't matter which comes first and that's why I am calling distinct to keep just one entry per block.

This was me mostly exercising the functional programming muscle instead of doing it imperatively and maybe more legible.